sycl/doc/design/IVDEPAndParallelLoopAnnotations.md

```diff
@@ -0,0 +1,195 @@
+# Implementation design for parallel loop annotations and `ivdep`
+
+This document describes the LLVM and SPIR-V implementation for representing
+loop-carried memory dependence distance (or lack-there off) for kernel loops
+with parallelism annotations like ivdep.
+
+## SYCL Attributes
+
+### Attributes for specifying no loop-carried dependences: `[[intel::ivdep]]` and `[[intel::ivdep(array)]]`
+
+When placed on a kernel loop, these attributes indicate that there are no
+loop-carried memory (data) dependences on all pointer accesses (`ivdep`) or on the specified
+arrays (`ivdep(array)`). This annotation allows the compiler to avoid making
+conservative assumptions when it cannot infer the loop-carried dependences on a
+specific array. For example consider the following kernel code:
+
+```
+q.single_task([=] {
+    for (int i = 0; i < N; ++i) {
+        B[i] += B[idx[i]];
+    }
+}).wait();
+```
+
+Since the access `B[idx[i]]` is unknown at compile time, the compiler will have
+to make the conservative assumption that there is a loop carried dependence of
+distance 1 between the write to B and the read from B (in both directions) when
+it tries to parallelize the loop. If however `ivdep` is used, then the
+annotation tells the compiler that there are no loop-carried dependences on `B`,
+and the compiler can make parallelism decisions without this assumption.
+
+```
+q.single_task([=] {
+    [[intel::ivdep]]
+    for (int i = 0; i < N; ++i) {
+        B[i] += B[idx[i]];
+    }
+}).wait();
+```
+
+### Attributes for specifying loop-carried dependence distance: `[[intel::ivdep(safelen)]]` and `[[intel::ivdep(array,safelen)]]`
+
+When placed on a kernel loop, these attributes indicate that any loop-carried
+dependences on all pointer accesses (`ivdep(safelen)`) or on the specified
+arrays/pointers (`ivdep(array,safelen)`) have a distance of at least `safelen`.
+This can be used to disambiguate any aliasing when the compiler cannot
+automatically infer the dependence distance and instead has to make the
+conservative assumption that the distance is 1.
+
+### Marking parallel loops
```

Comment:
This section feels out-of-place since we aren't proposing any design at the SYCL level at this time.  I'd maybe add more of a preamble to it explaining that there is no such feature planned at this time but we're mentioning this for completeness to contrast it with ivdep and because there is an existing LLVM IR representation with these semantics.